src/compiler/preprocess/index.ts

```diff
@@ -222,98 +172,38 @@ export default async function preprocess(
 	const style = preprocessors.map(p => p.style).filter(Boolean);
 
-	// sourcemap_list is sorted in reverse order from last map (index 0) to first map (index -1)
-	// so we use sourcemap_list.unshift() to add new maps
-	// https://github.com/ampproject/remapping#multiple-transformations-of-a-file
-	const sourcemap_list: Array<DecodedSourceMap | RawSourceMap> = [];
-
-	// TODO keep track: what preprocessor generated what sourcemap? to make debugging easier = detect low-resolution sourcemaps in fn combine_mappings
+	const result = new PreprocessResult(source, filename);
 
-	for (const fn of markup) {
+	// TODO keep track: what preprocessor generated what sourcemap?
+	// to make debugging easier = detect low-resolution sourcemaps in fn combine_mappings
 
-		// run markup preprocessor
-		const processed = await fn({
-			content: source,
+	for (const process of markup) {
+		const processed = await process({
+			content: result.source,
 			filename
 		});
 
 		if (!processed) continue;
 
-		if (processed.dependencies) dependencies.push(...processed.dependencies);
-		source = processed.code;
-		if (processed.map) {
-			sourcemap_list.unshift(
-				typeof(processed.map) === 'string'
+		result.update_source({
+			string: processed.code,
+			map: processed.map
+				? // TODO: can we use decode_sourcemap?
+				  typeof processed.map === 'string'
 					? JSON.parse(processed.map)
 					: processed.map
-			);
-		}
-	}
-
-	async function preprocess_tag_content(tag_name: 'style' | 'script', preprocessor: Preprocessor) {
-		const get_location = getLocator(source);
-		const tag_regex = tag_name === 'style'
-			? /<!--[^]*?-->|<style(\s[^]*?)?(?:>([^]*?)<\/style>|\/>)/gi
-			: /<!--[^]*?-->|<script(\s[^]*?)?(?:>([^]*?)<\/script>|\/>)/gi;
-
-		const res = await replace_async(
-			file_basename,
-			source,
-			get_location,
-			tag_regex,
-			async (match, attributes = '', content = '', offset) => {
-				const no_change = () => StringWithSourcemap.from_source(
-					file_basename, match, get_location(offset));
-				if (!attributes && !content) {
-					return no_change();
-				}
-				attributes = attributes || '';
-				content = content || '';
-
-				// run script preprocessor
-				const processed = await preprocessor({
-					content,
-					attributes: parse_attributes(attributes),
-					filename
-				});
-				if (processed && processed.dependencies) {
-					dependencies.push(...processed.dependencies);
-				}
-				if (!processed || !processed.map && processed.code === content) {
-					return no_change();
-				}
-				return get_replacement(file_basename, offset, get_location, content, processed, `<${tag_name}${attributes}>`, `</${tag_name}>`);
-			}
-		);
-		source = res.string;
-		sourcemap_list.unshift(res.map);
+				: undefined,
+			dependencies: processed.dependencies
+		});
 	}
 
-	for (const fn of script) {
-		await preprocess_tag_content('script', fn);
+	for (const process of script) {
+		result.update_source(await process_tag('script', process, result));
```

Comment:
since `process_tag` already taken in `result`, why not calling `result.update_source` within?

Reply:
I feel this makes clearer what is going on: `process_tag` and and `process_markup` are pure functions and you can clearly see in the calling code that the purpose is to update `result`. The fact that mutations were spread out through the code was actually the major issue I wanted to fix with this change.

(Note that the third parameter required by `process_tag` is actually not a `PreprocessResult` but a `Source`, which is a subset of it)